CURRENT VERSION: 0.45

********************************************************************************

Changes of v 0.44 to v 0.45 (February 2018)

* Review of parameter values when the mutation function is called in rsitesearch.py.

* Processing of verbose and trace options of simcasavaids.py.

********************************************************************************

Changes of v 0.43 to v 0.44 (January 2018)

* Enzymes with ambiguous restriction sites can be used.

* Parameters enzyme1 and enzyme2 can be equal in rsiteseach.py (in this case,
  it simulates a digestion with only one enzyme).

* simcasavaids.py is a new tool which fixes sequence identifiers of a FASTQ read
  file generated by ddRADseqTools to compatible format with CASAVA.

* Several minor improvements.

********************************************************************************

Changes of v 0.42 to v 0.43 (October 2017)

* Parameters enzyme1 and enzyme2 can be an identifier or a restriction site
  sequence, independently of each other.

********************************************************************************

Changes of v 0.40 to v 0.42 (July 2017)

* Fixed a bug in the algorithm to mutate sequences that occurs when there is
  more than one SNP or indel.
 
* Added new options in the programs to control the printing of verbose and
  development info, and the making of statistical graphics.

We thank to the professor Alex Buerkle his comments to improve ddRADseqTools.
